### PR TITLE
add core.sync.event to be used as Windows events

### PR DESCRIPTION
**Pull request by rainers**
extracted from https://github.com/dlang/druntime/pull/2514

### CODE REVIEW

File: src/core/sync/event.d [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+ *    $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0).
+ *    (See accompanying file LICENSE)
+ * Authors: Rainer Schuetze
+ * Source:    $(DRUNTIMESRC core/sync/event.d)
+ */
+module core.sync.event;
+
+version (Windows)
+{
+    import core.sys.windows.basetsd /+: HANDLE +/;
```
Comment (jacob-carlborg) [inline src/core/sync/event.d]: Why these comments?

File: src/core/sync/event.d [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+        {
+            pthread_mutex_lock(&m_mutex);
+            m_state = true;
+            for (auto waiter = m_waiter; waiter != null; waiter = waiter.next)
+            {
+                pthread_mutex_lock(&waiter.mutex);
+                pthread_cond_signal(&waiter.cond);
+                pthread_mutex_unlock(&waiter.mutex);
+            }
+            pthread_mutex_unlock(&m_mutex);
```
Comment (jacob-carlborg) [inline src/core/sync/event.d]: `scope(exit)`?

File: src/core/sync/event.d [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+            assert(!m_waiter);
+            pthread_mutex_destroy(&m_mutex) == 0 ||
+                abort("Error: pthread_mutex_destroy failed.");
+        }
+    }
+
+
+    /**
+     * Set the event to "signaled", so that waiting clients are resumed
+     */
```
Comment (jacob-carlborg) [inline src/core/sync/event.d]: Using `///` would save two lines.

File: src/core/sync/event.d [hunk 4]
```diff
@@ -0,0 +1,318 @@
+/**
+ * The event module provides a primitive for lightweight signalling of other threads
+ * (aka Windows Events)
```
Comment (jacob-carlborg) [inline src/core/sync/event.d]: Reading this, I though the implementation was for Windows only.

File: src/core/sync/event.d [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+else
+{
+    static assert(false, "Platform not supported");
+}
+
+import core.time;
+import core.internal.abort : abort;
+
+/**
+ * This class represents an event. Clients of an event are suspended while waiting
```
Comment (jacob-carlborg) [inline src/core/sync/event.d]: This is a struct.

File: src/core/sync/event.d [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+    }
+
+    /**
+     * Initializes an event object.
+     *
+     * Params:
+     *  manualReset  = the state of the event is not reset automatically after resuming waiting clients
+     *  initialState = initial state of the signal
+     */
+    void initialize(bool manualReset, bool initialState)
```
Comment (jacob-carlborg) [inline src/core/sync/event.d]: Why is this a separate method?

File: src/core/sync/event.d (hunk 1 above)
Comment (rainers) [reply src/core/sync/event.d]: This follows the example of https://github.com/dlang/druntime/pull/2400/files (selective imports create a public alias).

File: src/core/sync/event.d (hunk 2 above)
Comment (rainers) [reply src/core/sync/event.d]: The function is nothrow anyway, so it doesn't make a difference in  code generation. With a short function like this, having the symmetry show up in the function is slightly better IMO.

File: src/core/sync/event.d (hunk 3 above)
Comment (rainers) [reply src/core/sync/event.d]: Ok.

File: src/core/sync/event.d (hunk 5 above)
Comment (rainers) [reply src/core/sync/event.d]: Ok. It started out as a class, but no longer.

File: src/core/sync/event.d (hunk 6 above)
Comment (rainers) [reply src/core/sync/event.d]: You cannot pass arguments to the constructor if it is part of another struct, e.g. https://github.com/dlang/druntime/pull/2514/files#diff-6615611ba51118215a271e3500c61122R2605. You have to use an explicit `initialize`-call then. `__ctor` could be an alternative, but I think it should not be part of the API.

File: src/core/sync/event.d (hunk 1 above)
Comment (wilzbach) [reply src/core/sync/event.d]: Hasn't this been fixed a year ago or so?

File: src/core/sync/event.d (hunk 4 above)
Comment (rainers) [reply src/core/sync/event.d]: reworded slightly.

File: src/core/sync/event.d (hunk 1 above)
Comment (wilzbach) [reply src/core/sync/event.d]: Oh well, looks like it's really still in deprecation, but there's a PR to remove them for good:

https://github.com/dlang/dmd/pull/9393

File: src/core/sync/event.d (hunk 6 above)
Comment (jacob-carlborg) [reply src/core/sync/event.d]: Is there a downside to do `evStart = Event(...)` in `Gcx.initialize`?

Comment (rainers) [reply src/core/sync/event.d]: I've explicitly added `@disable this(this)` as this could duplicate the handle reference.

Comment (jacob-carlborg) [reply src/core/sync/event.d]: Does not seem to prevent assignment: https://run.dlang.io/is/iefkBG.

Comment (rainers) [reply src/core/sync/event.d]: Hmm, seems like opAssign should be disabled, too.

Comment (jacob-carlborg) [reply src/core/sync/event.d]: If that's required then I'm fine with using `initialize` instead of a constructor.

File: src/core/sync/event.d [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+    }
+
+    @disable this(this);
+
+    ~this()
+    {
+        terminate();
+    }
+
+    void terminate()
```
Comment (jacob-carlborg) [inline src/core/sync/event.d]: No documentation.

**Comment by jacob-carlborg** [comment]
No changelog.

File: src/core/sync/event.d (hunk 1 above)
Comment (thewilsonator) [reply src/core/sync/event.d]: That will go in very soon.

**Comment by rainers** [comment]
Added changelog and a bit more documentation. I've also added a bool to keep track of the initialization state of the posix version.

>     * uncommented selective imports

I think this should be done for all the imports once the dmd deprecation is gone (https://github.com/dlang/dmd/pull/9393 doesn't help for this case yet).

**Comment by thewilsonator** [comment]
`posix.mak:348: recipe for target 'checkwhitespace' failed`

File: changelog/core_sync_event.dd [hunk 8]
```diff
@@ -0,0 +1,34 @@
+New module core.sync.event
+
+This module provides a primitive for lightweight signaling of other threads
+by emulating Windows events on posix. It can be used to start execution of
```
Comment (jacob-carlborg) [inline changelog/core_sync_event.dd]: * Perhaps I'm picky, but now it sounds like it's for Posix only. I would write something like: "... a cross-platform interface  for lightweight signaling of other threads. "
* `posix` -> `Posix`

Comment (rainers) [reply changelog/core_sync_event.dd]: No problem ;-) Amended.

File: changelog/core_sync_event.dd [hunk 9]
```diff
@@ -0,0 +1,33 @@
+New module core.sync.event
+
+This module provides a cross-platform interface for lightweight signaling of other threads.
+It can be used to start execution of multiple waiting threads simultaniously.
```
Comment (kinke) [inline changelog/core_sync_event.dd]: simultan**e**ously

File: src/core/sync/event.d (hunk 1 above)
Comment (jacob-carlborg) [reply src/core/sync/event.d]: @rainers this has been merged now.

Comment (rainers) [reply src/core/sync/event.d]: Not good  enough:
```
test.d(3): Deprecation: core.sync.event.HANDLE is not visible from module test
```

Comment (jacob-carlborg) [reply src/core/sync/event.d]: 😞

**Comment by kinke** [comment]
Rainer, have you explored whether [pthread_cond_broadcast](http://pubs.opengroup.org/onlinepubs/009696699/functions/pthread_cond_broadcast.html) and a single condition variable could work too?

**Comment by rainers** [comment]
> Rainer, have you explored whether [pthread_cond_broadcast](http://pubs.opengroup.org/onlinepubs/009696699/functions/pthread_cond_broadcast.html) and a single condition variable could work too?

Didn't know about it, now replaced the manual implementation.

File: src/core/sync/event.d [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+    {
+        pthread_mutex_t m_mutex;
+        pthread_cond_t m_cond;
+        bool m_initalized;
+        bool m_state;
+        bool m_manualReset;
+    }
+}
+
+// Test single-thread (non-shared) use.
```
Comment (wilzbach) [inline src/core/sync/event.d]: Maybe make this example public, s.t. there's at least one user-facing example.

Comment (rainers) [reply src/core/sync/event.d]: I think this example is a pretty bad one as it doesn't involve multiple threads. Maybe better to move the example from the changelog into the documentation?

Comment (wilzbach) [reply src/core/sync/event.d]: Sure. I just saw that the module is currently without any public example.

File: src/core/sync/event.d [hunk 11]
```diff
@@ -37,6 +37,35 @@ import core.internal.abort : abort;
...
+        for (int i = 0; i < 10; ++i)
+            group.create(&doProcess);
+
+        buffer = std.file.read(filename);
+        event.set();
+        group.joinAll();
+        event.terminate();
+    }
+}
+---
```
Comment (Geod24) [inline src/core/sync/event.d]: Why isn't it a documented unittest instead ?

Comment (rainers) [reply src/core/sync/event.d]: I think its already a bit too long for a really good example, and making it actually runnable would make it even longer.

**Pull request merged**